python/singa/sonnx.py

```diff
@@ -0,0 +1,242 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+
+from __future__ import division
+import pickle
+from singa import tensor
+from singa.tensor import Tensor
+from singa import autograd
+from singa import optimizer
+import numpy as np
+import onnx
+from onnx import numpy_helper
+
+from collections import Counter, deque
+import numpy as np
+import math
+
+from .tensor import Tensor
+from . import layer
+from singa.proto import model_pb2
+from . import singa_wrap as singa
+#from .tensor import einsum
+from autograd import *
+
+def load_onnx_model(name = 'singonnx.pkl'):
+    '''
+        load onnx model
+        Args:
+            name:name of onnx model file
+        Return:
+            onnx model
+    '''
+    with open(name, 'rb') as f:
+        model = pickle.load(f)
+    return model
+
+def onnx_model_init(inputs,model):
+    '''
+    init onnx model
+    Args:
+        inputs: input data for model
+        model: onnx model
+    Return:
+        a dictionary for save all of the node of singa computation graph
+    '''
+    a = {}
+    a['X'] = inputs
+    for i in model.graph.node:
+        if (i.op_type == 'Constant'):
+            a[str(i.output[0])] = tensor.from_numpy(numpy_helper.to_array(i.attribute[0].t))
+            a[str(i.output[0])].stores_grad = True
+    return a
+
+def onnx_loss(a,model,target):
+    '''
+    get onnx model loss
+    Args:
+        a: singa computational graph nodes dictionary
+        model: onnx model
+        target:
+    Return:
+        loss for onnx model
+    '''
+    for i in model.graph.node:
+        if (i.op_type == 'Constant'):
+            pass
+            # do nothing
+        if (i.op_type == 'LeakyRelu'):
+            a[str(i.output[0])] = autograd.relu(a[str(i.input[0])])
+        elif (i.op_type == 'Softmax'):
+            a[str(i.output[0])] = autograd.softmax(a[str(i.input[0])])
+        elif (i.op_type == 'Add'):
+            if(str(i.input[1])[-1] == 'b'):
+                a[str(i.output[0])] = autograd.add_bias(a[str(i.input[0])], a[str(i.input[1])])
+            else:
+                a[str(i.output[0])] = autograd.add(a[str(i.input[0])],a[str(i.input[1])])
+        elif (i.op_type == 'MatMul'):
+            a[str(i.output[0])] = autograd.matmul(a[str(i.input[0])], a[str(i.input[1])])
+
+    loss = autograd.cross_entropy(a['Y'], target)
+    return loss
+
+
+
+def get_onnx_model(y, dy=None):
+    '''
+    from singa computational graph(loss) to get onnx model
+    Args:
+        y:loss of singa autograd
+    Return:
+        onnx model
+    '''
+    ######################
+    import onnx
+    from onnx import helper
+    from onnx import AttributeProto, TensorProto, GraphProto
+    from onnx import numpy_helper
+    import numpy as np
+    from singa import tensor
+    X = helper.make_tensor_value_info('X', TensorProto.FLOAT, [1, 2])
```

Comment:
what does the hardcode number [1, 2] mean?

Reply:
It's the input size.
I just checked that if the input size is fit for the model structure, the result is right.
This number can be ignored.